cmd/oras/internal/display/metadata/json/repo_list.go

```diff
@@ -0,0 +1,49 @@
+/*
+Copyright The ORAS Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package json
+
+import (
+	"io"
+
+	"oras.land/oras/cmd/oras/internal/display/metadata"
+	"oras.land/oras/cmd/oras/internal/display/metadata/model"
+	"oras.land/oras/cmd/oras/internal/output"
+)
+
+// repoListHandler handles JSON metadata output for repo ls command.
+type repoListHandler struct {
+	out   io.Writer
+	model *model.Repositories
+}
+
+// NewRepoListHandler creates a new handler for repo ls events.
+func NewRepoListHandler(out io.Writer) metadata.RepoListHandler {
+	return &repoListHandler{
+		out:   out,
+		model: model.NewRepositories(),
+	}
+}
+
+// OnRepositoryListed implements metadata.RepoListHandler.
+func (h *repoListHandler) OnRepositoryListed(repo string) error {
+	h.model.AddRepository(repo)
```

Comment:
This pattern is going to be slow on very large lists. It would be pretty trivial to generate json for repos and tags, but future problem I guess.

Reply:
>  It would be pretty trivial to generate json for repos and tags

Yeah indeed, but printing the JSON struct line-by-line doesn't form a valid JSON object until the end. For example:

```json
{
"repositories": [
    "foo"
    "bar"


But we might want to limit the max number of repos/tags to avoid OOM